File: src/world/worldgenerator.ts
// Port of Core/worldgenerator.{h,cpp}.
// The RNG call order matches the original exactly so a seed reproduces the same
// map (paired with the MSVCRT rand() in core/rng.ts). std::vector::at() throwing
// on out-of-range is reproduced with at()/setAt() helpers so the C++ try/catch
// control flow (and therefore the rand() consumption) is preserved.

import { rand, srand } from '../core/rng';
import { Coordinate } from '../core/coordinate';
import { ImageVectors, AnimationOptions, AnimationOption, ImageVector } from '../core/images';
import { TileBase } from '../model/tile';
import { Grassland, Forest, AbundantForest, Mountain, River } from '../model/tiles';
import { Mikontalo } from '../model/building';
import { IGameEventHandler, IObjectManager, IGameScene, IGameSettingsManager } from '../model/base';

class RangeErr extends Error {}

function at(matrix: number[][], y: number, x: number): number {
  if (y < 0 || y >= matrix.length) throw new RangeErr('y');
  const row = matrix[y];
  if (x < 0 || x >= row.length) throw new RangeErr('x');
  return row[x];
}
function setAt(matrix: number[][], y: number, x: number, value: number): void {
  if (y < 0 || y >= matrix.length) throw new RangeErr('y');
  const row = matrix[y];
  if (x < 0 || x >= row.length) throw new RangeErr('x');
  row[x] = value;
}

export interface WorldGenDeps {
  objectManager: IObjectManager & { addTiles(tiles: TileBase[]): void };
  eventHandler: IGameEventHandler;
  gameSettings: IGameSettingsManager;
  scene: IGameScene;
}

export class WorldGenerator {
  generateMap(sizeX: number, sizeY: number, seed: number, deps: WorldGenDeps): void {
    srand(seed);

    const tiles: TileBase[] = [];
    const matrix = this.generateTerrain(sizeX, sizeY);

    for (let x = 0; x < sizeX; x++) {
      for (let y = 0; y < sizeY; y++) {
        const val = matrix[y][x];
        let imageVector: ImageVector = ImageVectors.GRASSLAND;
        let animationOption: AnimationOption = AnimationOptions.GRASSLAND;
        let spawnMikontalo = false;
        let tile: TileBase;

        const coord = new Coordinate(x, y);
        const eh = deps.eventHandler;
        const om = deps.objectManager;

        if (val === 0) {
          tile = new Grassland(coord, 1, 1, eh, om);
          imageVector = ImageVectors.GRASSLAND;
          animationOption = AnimationOptions.GRASSLAND;
        } else if (val === 1) {
          tile = new Forest(coord, 1, 1, eh, om);
          const rnd = (rand() % 2) - 1;
          imageVector = rnd === 0 ? ImageVectors.FOREST_1 : ImageVectors.FOREST_2;
          animationOption = AnimationOptions.FOREST;
        } else if (val === 2) {
          tile = new Mountain(coord, 1, 1, eh, om);
          imageVector = ImageVectors.MOUNTAIN;
          animationOption = AnimationOptions.MOUNTAIN;
        } else if (val === 4) {
          tile = new River(coord, 1, 1, eh, om);
          imageVector = this.getRiverOrientation(x, y, matrix, 4)[0];
          animationOption = AnimationOptions.RIVER;
        } else if (val === 5) {
          tile = new Mountain(coord, 1, 1, eh, om);
          imageVector = ImageVectors.MOUNTAIN_FOREST;
          animationOption = AnimationOptions.MOUNTAIN_FOREST;
        } else if (val === 3) {
          tile = new Grassland(coord, 1, 1, eh, om);
          imageVector = ImageVectors.GRASSLAND;
          animationOption = AnimationOptions.GRASSLAND;
          spawnMikontalo = true;
        } else if (val === 6) {
          tile = new AbundantForest(coord, 1, 1, eh, om);
          imageVector = ImageVectors.ABUNDANT_FOREST;
          animationOption = AnimationOptions.FOREST;
        } else {
          tile = new Grassland(coord, 1, 1, eh, om);
          imageVector = ImageVectors.GRASSLAND;
          animationOption = AnimationOptions.GRASSLAND;
        }

        tile.setGameSettings(deps.gameSettings);

        if (spawnMikontalo) {
          const mikontalo = new Mikontalo(eh, om, null);
          mikontalo.setImageFiles(ImageVectors.MIKONTALO);
          tile.addBuilding(mikontalo);
          eh.updateTile(tile);
        }

        if (tile instanceof River) {
          tile.setRiverOrientation(this.getRiverOrientation(x, y, matrix, 4)[1]);
          if (imageVector === ImageVectors.RIVER_EW) tile.setRiverShape('EW');
          if (imageVector === ImageVectors.RIVER_NS) tile.setRiverShape('NS');
          if (imageVector === ImageVectors.RIVER_NE) tile.setRiverShape('NE');
          if (imageVector === ImageVectors.RIVER_NW) tile.setRiverShape('NW');
          if (imageVector === ImageVectors.RIVER_SW) tile.setRiverShape('SW');
          if (imageVector === ImageVectors.RIVER_SE) tile.setRiverShape('SE');
        }

        tile.setImageFiles(imageVector);
        tile.setAnimationOption(animationOption);
        deps.scene.drawItem(tile);
        tiles.push(tile);
      }
    }

    deps.objectManager.addTiles(tiles);
  }

  generateTerrain(sizeX: number, sizeY: number): number[][] {
    const matrix: number[][] = [];

    // Phase 1: 15% forest seed.
    for (let y = 0; y < sizeY; y++) {
      const row: number[] = [];
      for (let x = 0; x < sizeX; x++) {
        const rnd = (rand() % 100) + 1;
        row.push(rnd < 15 ? 1 : 0);
      }
      matrix.push(row);
    }

    // Phase 2: cluster forests (60% spread to 8 neighbours; faithful to the
    // try/catch that aborts a cell's remaining spreads once an .at() throws).
    const tempMatrix = matrix.map((r) => [...r]);
    for (let y = 0; y < sizeY; y++) {
      for (let x = 0; x < sizeX; x++) {
        if (tempMatrix[y][x] === 1) {
          try {
            if ((rand() % 100) + 1 > 40) setAt(matrix, y - 1, x - 1, 1);
            if ((rand() % 100) + 1 > 40) setAt(matrix, y - 1, x, 1);
            if ((rand() % 100) + 1 > 40) setAt(matrix, y - 1, x + 1, 1);
            if ((rand() % 100) + 1 > 40) setAt(matrix, y, x - 1, 1);
            if ((rand() % 100) + 1 > 40) setAt(matrix, y, x + 1, 1);
            if ((rand() % 100) + 1 > 40) setAt(matrix, y + 1, x - 1, 1);
            if ((rand() % 100) + 1 > 40) setAt(matrix, y + 1, x, 1);
            if ((rand() % 100) + 1 > 40) setAt(matrix, y + 1, x + 1, 1);
          } catch {
            continue;
          }
        }
      }
    }

    // Phase 3: a single winding river.
    const dir = (rand() % 2) - 1; // 0 or -1
    const startingTileX = (rand() % (sizeX - 4)) + 2;
    const startingTileY = (rand() % (sizeY - 4)) + 2;
    let currentX = 0;
    let currentY = 0;
    const lastDirs = [0, 0];
    if (dir === 0) currentX = startingTileX;
    else currentY = startingTileY;

    for (;;) {
      if (currentX >= sizeX || currentX < 0) break;
      if (currentY >= sizeY || currentY < 0) break;
      matrix[currentY][currentX] = 4;

      let nextDir: number;
      const back = lastDirs[lastDirs.length - 1];
      const prev = lastDirs[lastDirs.length - 2];
      if (back === 0 && prev === 0) nextDir = rand() % 3;
      else if (back === 0 && prev === 1) nextDir = (rand() % 2) - 1;
      else if (back === 0 && prev === 2) nextDir = ((rand() % 2) - 1) * 2;
      else if (back === 1) nextDir = (rand() % 2) - 1;
      else if (back === 2) nextDir = ((rand() % 2) - 1) * 2;
      else nextDir = 0;

      if (dir === 0) {
        if (nextDir === 0) currentY += 1;
        if (nextDir === 1) currentX += 1;
        if (nextDir === 2) currentX -= 1;
      } else {
        if (nextDir === 0) currentX += 1;
        if (nextDir === 1) currentY += 1;
        if (nextDir === 2) currentY -= 1;
      }
      lastDirs.push(nextDir);
    }

    // Phase 4: mountains (loop bound re-evaluated every iteration, as in C++).
    for (let i = 0; i < (rand() % Math.round(sizeX * sizeY * 0.3)) + 4; i++) {
      const rndX = rand() % sizeX;
      const rndY = rand() % sizeY;
      if (matrix[rndY][rndX] === 1) matrix[rndY][rndX] = 5;
      else if (matrix[rndY][rndX] === 0) matrix[rndY][rndX] = 2;
      else continue;
    }

    // Phase 5: abundant forests.
    for (let x = 0; x < Math.trunc((sizeX * sizeY) / 30); x++) {
      const rndX = rand() % sizeX;
      const rndY = rand() % sizeY;
      if (matrix[rndY][rndX] === 4) continue;
      matrix[rndY][rndX] = 6;
    }

    // Phase 6: a single Mikontalo.
    for (;;) {
      const rndX = rand() % sizeX;
      const rndY = rand() % sizeY;
      if (matrix[rndY][rndX] === 4) continue;
      matrix[rndY][rndX] = 3;
      break;
    }

    return matrix;
  }

  getRiverOrientation(x: number, y: number, matrix: number[][], num: number): [ImageVector, number] {
    const sizeX = matrix[0].length;
    const sizeY = matrix.length;
    try {
      if (at(matrix, y - 1, x) === num && at(matrix, y, x - 1) === num) return [ImageVectors.RIVER_NW, 3];
      if (at(matrix, y - 1, x) === num && at(matrix, y, x + 1) === num) return [ImageVectors.RIVER_NE, 3];
      if (at(matrix, y + 1, x) === num && at(matrix, y, x - 1) === num) return [ImageVectors.RIVER_SW, 3];
      if (at(matrix, y + 1, x) === num && at(matrix, y, x + 1) === num) return [ImageVectors.RIVER_SE, 3];
      if (at(matrix, y - 1, x) === num && at(matrix, y + 1, x) === num) return [ImageVectors.RIVER_NS, 1];
      if (at(matrix, y, x - 1) === num && at(matrix, y, x + 1) === num) return [ImageVectors.RIVER_EW, 0];
    } catch {
      try {
        if (x === 0 && y === 0) {
          if (at(matrix, y + 1, x) === num) return [ImageVectors.RIVER_NS, 1];
          if (at(matrix, y, x + 1) === num) return [ImageVectors.RIVER_EW, 0];
        } else if (x === 0 && y === sizeY - 1) {
          if (at(matrix, y - 1, x) === num) return [ImageVectors.RIVER_NS, 1];
          if (at(matrix, y, x + 1) === num) return [ImageVectors.RIVER_EW, 0];
        } else if (x === sizeX - 1 && y === 0) {
          if (at(matrix, y + 1, x) === num) return [ImageVectors.RIVER_NS, 1];
          if (at(matrix, y, x - 1) === num) return [ImageVectors.RIVER_EW, 0];
        } else if (x === sizeX - 1 && y === sizeY - 1) {
          if (at(matrix, y - 1, x) === num) return [ImageVectors.RIVER_NS, 1];
          if (at(matrix, y, x - 1) === num) return [ImageVectors.RIVER_EW, 0];
        } else if (y === 0) {
          if (at(matrix, y, x - 1) === num && at(matrix, y, x + 1) === num) return [ImageVectors.RIVER_EW, 0];
          else if (at(matrix, y + 1, x) === num && at(matrix, y, x - 1) === num) return [ImageVectors.RIVER_SW, 3];
          else if (at(matrix, y + 1, x) === num && at(matrix, y, x + 1) === num) return [ImageVectors.RIVER_SE, 3];
          else if (at(matrix, y, x - 1) === num) return [ImageVectors.RIVER_NW, 3];
          else if (at(matrix, y, x + 1) === num) return [ImageVectors.RIVER_NE, 3];
          else return [ImageVectors.RIVER_NS, 1];
        } else if (y === sizeY - 1) {
          if (at(matrix, y, x - 1) === num && at(matrix, y, x + 1) === num) return [ImageVectors.RIVER_EW, 0];
          else if (at(matrix, y - 1, x) === num && at(matrix, y, x - 1) === num) return [ImageVectors.RIVER_NW, 3];
          else if (at(matrix, y - 1, x) === num && at(matrix, y, x + 1) === num) return [ImageVectors.RIVER_NE, 3];
          else if (at(matrix, y, x - 1) === num) return [ImageVectors.RIVER_SW, 3];
          else if (at(matrix, y, x + 1) === num) return [ImageVectors.RIVER_SE, 3];
          else return [ImageVectors.RIVER_NS, 1];
        } else if (x === 0) {
          if (at(matrix, y - 1, x) === num && at(matrix, y + 1, x) === num) return [ImageVectors.RIVER_NS, 1];
          else if (at(matrix, y, x + 1) === num && at(matrix, y - 1, x) === num) return [ImageVectors.RIVER_NE, 3];
          else if (at(matrix, y, x + 1) === num && at(matrix, y + 1, x) === num) return [ImageVectors.RIVER_SE, 3];
          else if (at(matrix, y - 1, x) === num) return [ImageVectors.RIVER_NW, 3];
          else if (at(matrix, y + 1, x) === num) return [ImageVectors.RIVER_SW, 3];
          else return [ImageVectors.RIVER_EW, 0];
        } else if (x === sizeX - 1) {
          if (at(matrix, y - 1, x) === num && at(matrix, y + 1, x) === num) return [ImageVectors.RIVER_NS, 1];
          else if (at(matrix, y, x - 1) === num && at(matrix, y - 1, x) === num) return [ImageVectors.RIVER_NW, 3];
          else if (at(matrix, y, x - 1) === num && at(matrix, y + 1, x) === num) return [ImageVectors.RIVER_SW, 3];
          else if (at(matrix, y - 1, x) === num) return [ImageVectors.RIVER_NE, 3];
          else if (at(matrix, y + 1, x) === num) return [ImageVectors.RIVER_SE, 3];
          else return [ImageVectors.RIVER_EW, 0];
        }
      } catch {
        // matches the qDebug("Error with river orientation") branch
      }
    }
    // Original returns an empty tuple here; default to a straight N-S piece.
    return [ImageVectors.RIVER_NS, 1];
  }
}

File: reference/Game/Core/worldgenerator.cpp
/*
######################################################################
# TIE-02402 Ohjelmointi 3: Perusteet, S2019                          #
#                                                                    #
# Project: Colonizing Pirkanmaa                                      #
# Program description: Program instructions are located in           #
#                      Documentation/documentation.pdf               #
#                                                                    #
# File: worldgenerator.cpp, see worldgenerator.h for more info       #
#                                                                    #
# Authors: Otto Ranta-Ojala, 253561, otto.ranta-ojala@tuni.fi        #
#          Tino Kaartovuori, 254987, tino.kaartovuori@tuni.fi        #
######################################################################
*/


#include "Core/worldgenerator.h"

#include "Tiles/forest.h"
#include "Tiles/grassland.h"
#include "Tiles/mountain.h"
#include "Buildings/mikontalo.h"
#include "Tiles/river.h"
#include "Tiles/abundantforest.h"

#include "Graphics/imagevectors.h"

#include <QDebug>
#include <vector>
#include <iostream>
#include <math.h>
#include <tuple>


namespace Student {
WorldGenerator& WorldGenerator::getInstance()
{
    static WorldGenerator instance;
    return instance;
}

void WorldGenerator::generateMap(unsigned int size_x_,
        unsigned int size_y_,
        unsigned int seed,
        const std::weak_ptr<Course::iObjectManager> &objectmanager,
        const std::weak_ptr<Course::iGameEventHandler> &eventhandler,
        const std::weak_ptr<GameSettingsManager> &gamesettingsmanager,
        std::weak_ptr<GameScene> scene)
{
    this->addConstructor<Course::Forest>("Forest");
    this->addConstructor<Course::Grassland>("Grassland");
    this->addConstructor<Student::Mountain>("Mountain");
    this->addConstructor<Student::AbundantForest>("AbundantForest");
    this->addConstructor<Student::River>("River");

    srand(seed);

    std::vector<std::shared_ptr<Course::TileBase>> tiles;

    std::vector<std::vector<int>> matrix = generateTerrain(size_x_, size_y_);

    for (unsigned int x = 0; x < size_x_; ++x)
    {
        for (unsigned int y = 0; y < size_y_; ++y)
        {
            TileConstructorPointer ctor;
            std::vector<std::string> imageVector;
            Student::AnimationOption animationOption;

            bool spawnMikontalo = false;

            if (matrix.at(y).at(x) == 0) {
                ctor = m_ctors.find("Grassland")->second;
                imageVector = ImageVectors::GRASSLAND;
                animationOption = AnimationOptions::GRASSLAND;
            }
            else if (matrix.at(y).at(x) == 1) {
                ctor = m_ctors.find("Forest")->second;
                int rnd = (rand() % 2) - 1;
                if (rnd == 0) {
                    imageVector = ImageVectors::FOREST_1;
                }
                else {
                    imageVector = ImageVectors::FOREST_2;
                }
                animationOption = AnimationOptions::FOREST;
            }
            else if (matrix.at(y).at(x) == 2) {
                ctor = m_ctors.find("Mountain")->second;
                imageVector = ImageVectors::MOUNTAIN;
                animationOption = AnimationOptions::MOUNTAIN;
            }
            else if (matrix.at(y).at(x) == 4) {
                ctor = m_ctors.find("River")->second;
                imageVector = std::get<0>(getRiverOrientation(x, y, matrix, 4));

                animationOption = AnimationOptions::RIVER;
            }
            else if (matrix.at(y).at(x) == 5) {
                ctor = m_ctors.find("Mountain")->second;
                imageVector = ImageVectors::MOUNTAIN_FOREST;
                animationOption = AnimationOptions::MOUNTAIN_FOREST;
            }
            else if (matrix.at(y).at(x) == 3) {
                // Spawn Mikontalo to tile too.
                ctor = m_ctors.find("Grassland")->second;
                imageVector = ImageVectors::GRASSLAND;
                animationOption = AnimationOptions::GRASSLAND;
                spawnMikontalo = true;
            }
            else if (matrix.at(y).at(x) == 6) {
                ctor = m_ctors.find("AbundantForest")->second;
                imageVector = ImageVectors::ABUNDANT_FOREST;
                animationOption = AnimationOptions::FOREST;
            }
            else {
                ctor = m_ctors.find("Grassland")->second;
                imageVector = ImageVectors::GRASSLAND;
                animationOption = AnimationOptions::GRASSLAND;
            }

            tiles.push_back(ctor(Course::Coordinate(x, y), 1, 1,
                                 eventhandler.lock(), objectmanager.lock()));

            if (spawnMikontalo) {

                std::shared_ptr<Student::Mikontalo> mikontalo = std::make_shared<Student::Mikontalo>(
                            eventhandler.lock(),
                            objectmanager.lock(),
                            nullptr
                            );

                mikontalo->setImageFiles(ImageVectors::MIKONTALO);
                tiles.back()->addBuilding(mikontalo);
                eventhandler.lock()->updateTile(tiles.back());

            }

            if (std::dynamic_pointer_cast<Student::River>(tiles.back()) != nullptr) {
                std::dynamic_pointer_cast<Student::River>(tiles.back())->setRiverOrientation(std::get<1>(getRiverOrientation(x, y, matrix, 4)));
                if (imageVector == ImageVectors::RIVER_EW) {
                    std::dynamic_pointer_cast<Student::River>(tiles.back())->setRiverShape("EW");
                }
                if (imageVector == ImageVectors::RIVER_NS) {
                    std::dynamic_pointer_cast<Student::River>(tiles.back())->setRiverShape("NS");
                }
                if (imageVector == ImageVectors::RIVER_NE) {
                    std::dynamic_pointer_cast<Student::River>(tiles.back())->setRiverShape("NE");
                }
                if (imageVector == ImageVectors::RIVER_NW) {
                    std::dynamic_pointer_cast<Student::River>(tiles.back())->setRiverShape("NW");
                }
                if (imageVector == ImageVectors::RIVER_SW) {
                    std::dynamic_pointer_cast<Student::River>(tiles.back())->setRiverShape("SW");
                }
                if (imageVector == ImageVectors::RIVER_SE) {
                    std::dynamic_pointer_cast<Student::River>(tiles.back())->setRiverShape("SE");
                }
            }

            tiles.back()->setGameSettings(gamesettingsmanager.lock());
            tiles.back()->setImageFiles(imageVector);
            tiles.back()->setAnimationOption(animationOption);
            scene.lock()->drawItem(tiles.back());
        }
    }

    objectmanager.lock()->addTiles(tiles);

}

std::vector<std::vector<int>> WorldGenerator::generateTerrain(int size_x_, int size_y_) {

    std::vector<std::vector<int>> matrix;
    std::vector<int> row;

    int rnd;
    for (int y = 0; y < size_y_; ++y)
    {
        for (int x = 0; x < size_x_; ++x)
        {
            rnd = rand() % 100 + 1;

            if (rnd < 15) {
                row.push_back(1);
            }
            else {
                row.push_back(0);
            }
        }
        matrix.push_back(row);
        row.clear();
    }
    std::vector<std::vector<int>> temp_matrix;
    temp_matrix = matrix;
    for (int y = 0; y < size_y_; ++y)
    {
        for (int x = 0; x < size_x_; ++x)
        {
            if (temp_matrix.at(y).at(x) == 1) {

                try {
                    if (rand() % 100 + 1 > 40) {
                        matrix.at(y-1).at(x-1) = 1;
                    }
                    if (rand() % 100 + 1 > 40) {
                        matrix.at(y-1).at(x) = 1;
                    }
                    if (rand() % 100 + 1 > 40) {
                        matrix.at(y-1).at(x+1) = 1;
                    }
                    if (rand() % 100 + 1 > 40) {
                        matrix.at(y).at(x-1) = 1;
                    }
                    if (rand() % 100 + 1 > 40) {
                        matrix.at(y).at(x+1) = 1;
                    }
                    if (rand() % 100 + 1 > 40) {
                        matrix.at(y+1).at(x-1) = 1;
                    }
                    if (rand() % 100 + 1 > 40) {
                        matrix.at(y+1).at(x) = 1;
                    }
                    if (rand() % 100 + 1 > 40) {
                        matrix.at(y+1).at(x+1) = 1;
                    }
                }
                catch(...) {
                    continue;
                }

            }
        }
    }

    int dir = rand() % 2 - 1;
    int starting_tile_x = (rand() % (size_x_ - 4)) + 2;
    int starting_tile_y = (rand() % (size_y_ - 4)) + 2;
    int current_x = 0;
    int current_y = 0;

    int next_dir;

    std::vector<int> last_dirs = {0, 0};

    if (dir == 0) {
        current_x = starting_tile_x;
    } else {
        current_y = starting_tile_y;
    }

    while (true) {

        if (current_x >= size_x_ or current_x < 0) break;
        if (current_y >= size_y_ or current_y < 0) break;

        matrix.at(current_y).at(current_x) = 4;


        if (last_dirs.back() == 0 && last_dirs.at(last_dirs.size() - 2) == 0 ) {
            next_dir = rand() % 3;
        }
        else if (last_dirs.back() == 0 && last_dirs.at(last_dirs.size() - 2) == 1) {
            next_dir = (rand() % 2 - 1);
        }
        else if (last_dirs.back() == 0 && last_dirs.at(last_dirs.size() - 2) == 2) {
            next_dir = (rand() % 2 - 1) * 2;
        }
        else if (last_dirs.back() == 1) {
            next_dir = (rand() % 2 - 1);
        }
        else if (last_dirs.back() == 2) {
            next_dir = (rand() % 2 - 1) * 2;
        }
        else {
            next_dir = 0;
        }

        if (dir == 0) {
            if (next_dir == 0) {
                current_y += 1;
            }
            if (next_dir == 1) {
                current_x += 1;
            }
            if (next_dir == 2) {
                current_x -= 1;
            }
        } else {
            if (next_dir == 0) {
                current_x += 1;
            }
            if (next_dir == 1) {
                current_y += 1;
            }
            if (next_dir == 2) {
                current_y -= 1;
            }
        }

        last_dirs.push_back(next_dir);


    }

    for (int i = 0; i < (int)(rand() % (int)round(size_x_ * size_y_ * 0.30)) + 4; i++) {
        int rnd_x = rand() % size_x_;
        int rnd_y = rand() % size_y_;


        if (matrix.at(rnd_y).at(rnd_x) == 1) {
            matrix.at(rnd_y).at(rnd_x) = 5;
        }
        if (matrix.at(rnd_y).at(rnd_x) == 0) {
            matrix.at(rnd_y).at(rnd_x) = 2;
        }
        else {
            continue;
        }
    }

    for (int x = 0; x < (int)(size_x_*size_y_)/30; x++) {
        int rnd_x = rand() % size_x_;
        int rnd_y = rand() % size_y_ ;

        if (matrix.at(rnd_y).at(rnd_x) == 4) continue;

        matrix.at(rnd_y).at(rnd_x) = 6;
    }

    while (true) {
        int rnd_x = rand() % size_x_;
        int rnd_y = rand() % size_y_ ;

        if (matrix.at(rnd_y).at(rnd_x) == 4) continue;

        matrix.at(rnd_y).at(rnd_x) = 3;
        break;
    }

    return matrix;
}


std::tuple<std::vector<std::string>, int> WorldGenerator::getRiverOrientation
                     (int x_, int y_, std::vector<std::vector<int>> matrix_, int num)
{
    std::vector<std::vector<int>> matrix = matrix_;
    int x = x_;
    int y = y_;
    int size_x_ = matrix.at(0).size();
    int size_y_ = matrix.size();

    try {
        if (matrix.at(y-1).at(x) == num and matrix.at(y).at(x-1) == num) {
            return std::make_tuple(ImageVectors::RIVER_NW, 3);
        }
        if (matrix.at(y-1).at(x) == num and matrix.at(y).at(x+1) == num) {
            return std::make_tuple(ImageVectors::RIVER_NE, 3);
        }
        if (matrix.at(y+1).at(x) == num and matrix.at(y).at(x-1) == num) {
            return std::make_tuple(ImageVectors::RIVER_SW, 3);
        }
        if (matrix.at(y+1).at(x) == num and matrix.at(y).at(x+1) == num) {
            return std::make_tuple(ImageVectors::RIVER_SE, 3);
        }
        if (matrix.at(y-1).at(x) == num and matrix.at(y+1).at(x) == num) {
            return std::make_tuple(ImageVectors::RIVER_NS, 1);
        }
        if (matrix.at(y).at(x-1) == num and matrix.at(y).at(x+1) == num) {
            return std::make_tuple(ImageVectors::RIVER_EW, 0);
        }
    }
    catch (...) {
        try {
            if (x == 0 and y == 0) {
                if (matrix.at(y+1).at(x) == num) {
                    return std::make_tuple(ImageVectors::RIVER_NS, 1);
                }
                if (matrix.at(y).at(x+1) == num) {
                    return std::make_tuple(ImageVectors::RIVER_EW, 0);
                }
            }
            else if (x == 0 and y == size_y_ - 1) {
                if (matrix.at(y-1).at(x) == num) {
                    return std::make_tuple(ImageVectors::RIVER_NS, 1);
                }
                if (matrix.at(y).at(x+1) == num) {
                    return std::make_tuple(ImageVectors::RIVER_EW, 0);
                }
            }
            else if (x == size_x_ -1 and y == 0) {
                if (matrix.at(y+1).at(x) == num) {
                    return std::make_tuple(ImageVectors::RIVER_NS, 1);
                }
                if (matrix.at(y).at(x-1) == num) {
                    return std::make_tuple(ImageVectors::RIVER_EW, 0);
                }
            }
            else if (x == size_x_ - 1 and y == size_y_ - 1) {
                if (matrix.at(y-1).at(x) == num) {
                    return std::make_tuple(ImageVectors::RIVER_NS, 1);
                }
                if (matrix.at(y).at(x-1) == num) {
                    return std::make_tuple(ImageVectors::RIVER_EW, 0);
                }
            }
            else if (y == 0) {
                if (matrix.at(y).at(x-1) == num and matrix.at(y).at(x+1) == num) {
                    return std::make_tuple(ImageVectors::RIVER_EW, 0);
                }
                else if (matrix.at(y+1).at(x) == num and matrix.at(y).at(x-1) == num) {
                    return std::make_tuple(ImageVectors::RIVER_SW, 3);
                }
                else if (matrix.at(y+1).at(x) == num and matrix.at(y).at(x+1) == num) {
                    return std::make_tuple(ImageVectors::RIVER_SE, 3);
                }
                else if (matrix.at(y).at(x-1) == num) {
                    return std::make_tuple(ImageVectors::RIVER_NW, 3);
                }
                else if (matrix.at(y).at(x+1) == num) {
                    return std::make_tuple(ImageVectors::RIVER_NE, 3);
                }
                else {
                    return std::make_tuple(ImageVectors::RIVER_NS, 1);
                }
            }
            else if (y == size_y_ - 1) {
                if (matrix.at(y).at(x-1) == num and matrix.at(y).at(x+1) == num) {
                    return std::make_tuple(ImageVectors::RIVER_EW, 0);
                }
                else if (matrix.at(y-1).at(x) == num and matrix.at(y).at(x-1) == num) {
                    return std::make_tuple(ImageVectors::RIVER_NW, 3);
                }
                else if (matrix.at(y-1).at(x) == num and matrix.at(y).at(x+1) == num) {
                    return std::make_tuple(ImageVectors::RIVER_NE, 3);
                }
                else if (matrix.at(y).at(x-1) == num) {
                    return std::make_tuple(ImageVectors::RIVER_SW, 3);
                }
                else if (matrix.at(y).at(x+1) == num) {
                    return std::make_tuple(ImageVectors::RIVER_SE, 3);
                }
                else {
                    return std::make_tuple(ImageVectors::RIVER_NS, 1);
                }
            }
            else if (x == 0) {
                if (matrix.at(y-1).at(x) == num and matrix.at(y+1).at(x) == num) {
                    return std::make_tuple(ImageVectors::RIVER_NS, 1);
                }
                else if (matrix.at(y).at(x+1) == num and matrix.at(y-1).at(x) == num) {
                    return std::make_tuple(ImageVectors::RIVER_NE, 3);
                }
                else if (matrix.at(y).at(x+1) == num and matrix.at(y+1).at(x) == num) {
                    return std::make_tuple(ImageVectors::RIVER_SE, 3);
                }
                else if (matrix.at(y-1).at(x) == num) {
                    return std::make_tuple(ImageVectors::RIVER_NW, 3);
                }
                else if (matrix.at(y+1).at(x) == num) {
                    return std::make_tuple(ImageVectors::RIVER_SW, 3);
                }
                else {
                    return std::make_tuple(ImageVectors::RIVER_EW, 0);
                }
            }
            else if (x == size_x_ - 1) {
                if (matrix.at(y-1).at(x) == num and matrix.at(y+1).at(x) == num) {
                    return std::make_tuple(ImageVectors::RIVER_NS, 1);
                }
                else if (matrix.at(y).at(x-1) == num and matrix.at(y-1).at(x) == num) {
                    return std::make_tuple(ImageVectors::RIVER_NW, 3);
                }
                else if (matrix.at(y).at(x-1) == num and matrix.at(y+1).at(x) == num) {
                    return std::make_tuple(ImageVectors::RIVER_SW, 3);
                }
                else if (matrix.at(y-1).at(x) == num) {
                    return std::make_tuple(ImageVectors::RIVER_NE, 3);
                }
                else if (matrix.at(y+1).at(x) == num) {
                    return std::make_tuple(ImageVectors::RIVER_SE, 3);
                }
                else {
                    return std::make_tuple(ImageVectors::RIVER_EW, 0);
                }
            }
        }
        catch (...) {
            qDebug() << "Error with river orientation";
        }
    }

    return {};
}

} // namespace Course
